In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
dataset = pd.read_csv('Social_Network_Ads.csv')
dataset.head(15)

User ID  Gender  Age  EstimatedSalary  Purchased
0   15624510    Male   19            19000          0
1   15810944    Male   35            20000          0
2   15668575  Female   26            43000          0
3   15603246  Female   27            57000          0
4   15804002    Male   19            76000          0
5   15728773    Male   27            58000          0
6   15598044  Female   27            84000          0
7   15694829  Female   32           150000          1
8   15600575    Male   25            33000          0
9   15727311  Female   35            65000          0
10  15570769  Female   26            80000          0
11  15606274  Female   26            52000          0
12  15746139    Male   20            86000          0
13  15704987    Male   32            18000          0
14  15628972    Male   18            82000          0

In [2]:
X = dataset.iloc[:, 2:4].values.astype(np.float64)
y = dataset.iloc[:, 4].values.astype(np.float64)
pd.DataFrame(X).head(15)


0         1
0   19.0   19000.0
1   35.0   20000.0
2   26.0   43000.0
3   27.0   57000.0
4   19.0   76000.0
5   27.0   58000.0
6   27.0   84000.0
7   32.0  150000.0
8   25.0   33000.0
9   35.0   65000.0
10  26.0   80000.0
11  26.0   52000.0
12  20.0   86000.0
13  32.0   18000.0
14  18.0   82000.0

In [3]:
pd.DataFrame(y).head(15)

0
0   0.0
1   0.0
2   0.0
3   0.0
4   0.0
5   0.0
6   0.0
7   1.0
8   0.0
9   0.0
10  0.0
11  0.0
12  0.0
13  0.0
14  0.0

In [4]:
#Splitting
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =  train_test_split(X,y,test_size = 0.3, random_state = 54)

# Feature Scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
pd.DataFrame(X_test).head(15)


0         1
0   1.455306  0.053649
1  -1.276968 -1.378376
2  -0.052155 -0.530851
3   0.230494  0.053649
4  -0.617453 -1.056901
5   0.324710 -0.209376
6  -1.088535  0.316674
7  -0.711670 -1.553726
8   0.324710 -1.173801
9   0.890008 -1.056901
10  0.136277 -0.823101
11 -0.711670  0.287449
12 -0.240588 -1.466051
13  0.230494 -0.297051
14  0.042061  0.024424

In [5]:
def LR(X_train , y_train , rate , maxitr, coeff, m):
    X_train = np.append(X_train , np.matrix(np.ones(len(X_train[:,0]))).T , axis=1)
    for i in range(maxitr):
        lf = np.dot(X_train , coeff)
       
        hypothesis = 1/(1+np.exp(-lf))
        
        cost = -((y_train*np.log(hypothesis)) + ((1-y_train)*np.log(1-hypothesis)))/(2*m)
        gradiant = np.matmul((hypothesis.T-y_train),X_train)
        coeff = coeff - rate*gradiant.T
        #print(cost)
    return coeff
coeff = np.matrix('0.0;0.0;0.0')

coeff= LR(X_train , y_train ,1e-2 , 40000 ,    coeff , len(X_train[:,0]))
def predict(X_test , coeff):
    X_test = np.append(X_test , np.matrix(np.ones(len(X_test[:,0]))).T , axis=1)
    y_p= 1/(1+np.exp(np.dot(X_test, coeff)))

    for i in range(len(y_p[:,0])):
    
    
    
        if y_p[i,:] >=0.5:
            y_p[i,:] = 1
        else:
            y_p[i,:] = 0
    return y_p

In [6]:
y_predict = predict(X_test,coeff)
pd.DataFrame(y_predict)

0
0    0.0
1    1.0
2    1.0
3    1.0
4    1.0
5    1.0
6    1.0
7    1.0
8    1.0
9    1.0
10   1.0
11   1.0
12   1.0
13   1.0
14   1.0
15   0.0
16   0.0
17   1.0
18   0.0
19   0.0
20   0.0
21   1.0
22   1.0
23   1.0
24   0.0
25   1.0
26   1.0
27   0.0
28   1.0
29   1.0
..   ...
90   1.0
91   0.0
92   1.0
93   1.0
94   1.0
95   1.0
96   1.0
97   0.0
98   0.0
99   1.0
100  1.0
101  1.0
102  1.0
103  1.0
104  0.0
105  1.0
106  0.0
107  1.0
108  0.0
109  1.0
110  0.0
111  0.0
112  1.0
113  1.0
114  0.0
115  1.0
116  1.0
117  1.0
118  0.0
119  1.0

[120 rows x 1 columns]

In [7]:
pd.DataFrame(y_test).tail(30)

0
90   1.0
91   1.0
92   0.0
93   0.0
94   0.0
95   0.0
96   0.0
97   0.0
98   1.0
99   0.0
100  1.0
101  0.0
102  0.0
103  0.0
104  1.0
105  0.0
106  1.0
107  0.0
108  1.0
109  0.0
110  0.0
111  1.0
112  0.0
113  0.0
114  1.0
115  0.0
116  0.0
117  0.0
118  1.0
119  0.0

In [9]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_predict)

print(cm)


[[ 3 74]
 [33 10]]


confusion matrix 

Total correct predictions = 74+33 = 107

Total incorrect predictions = 11+2 = 13